holonote/annotate/table.py

```diff
@@ -192,34 +181,31 @@ def clear_edits(self, edit_type=None):
         self._edits = []
         self._index_mapping = {}
 
-    def add_annotation(self, regions, **fields):
+    def add_annotation(self, regions: dict[str, Any], spec: SpecDict, **fields):
         "Takes a list of regions or the special value 'annotation-regions' to use attached annotators"
         index_value = fields.pop(self._field_df.index.name)
-        if regions == 'annotator-regions':
-            regions = [a._region for a in self._annotators.values()]
-        if all(el=={} for el in regions):
-            print('No new region selected. Skipping')
-            return
         self._add_annotation_fields(index_value, fields=fields)
 
-        for region in regions:
-            if region == {}:
+        data = []
+        for kdim, value in regions.items():
+            if not value:
                 continue
 
-            region_type, dim1, dim2, value = (region['region_type'], region['dim1'],
-                                              region['dim2'], region['value'])
-            if region_type not in ['Range', 'Point']:
-                raise Exception('TODO: Currently only supporting Range and Point annotations')
+            d = {"region": spec[kdim]["region"], "dim": kdim, "value": value, "_id": index_value}
+            data.append(
+                # pd.DataFrame(d) does not work because tuples is expanded into multiple rows:
+                # pd.DataFrame({'v': (1, 2)})
+                pd.DataFrame(d.values(), index=d.keys()).T
+            )
 
-            new_region_row = pd.DataFrame([[region_type, dim1, dim2, value, index_value]], columns=self.columns)
-            self._region_df = pd.concat((self._region_df, new_region_row), ignore_index=True)
+        self._region_df = pd.concat((self._region_df, *data), ignore_index=True)
 
         self._edits.append({'operation':'insert', 'id':index_value})
         self._update_index()
 
-    def refresh_annotators(self, clear=False):
-        for annotator in self._annotators.values():
-            annotator.refresh(clear=clear)
+    # def refresh_annotators(self, clear=False):
+    #     for annotator in self._annotators.values():
+    #         annotator.refresh(clear=clear)
```

Comment:
These commented lines can now be deleted not there will be only a single annotator instance.

Reply:
Agree. I have just left this in to reduce the number of changes. I will follow up on this PR with the removal of this.